snuba/query/matcher.py

```diff
@@ -0,0 +1,201 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from datetime import date, datetime
+from typing import Generic, Mapping, Optional, Tuple, TypeVar
+
+from snuba.query.expressions import Column as ColumnExpr
+from snuba.query.expressions import Expression, FunctionCall
+
+TNode = TypeVar("TNode")
+
+
+@dataclass(frozen=True)
+class MatchResult:
+    """
+    Contains the parameters found in an expression by a Pattern.
+    Parameters are the equivalent to groups in regular expressions, they
+    are named parts of the expression we are matching that are identified
+    when a valid match is found.
+    """
+
+    expressions: Mapping[str, Expression] = field(default_factory=dict)
+    strings: Mapping[str, Optional[str]] = field(default_factory=dict)
+    ints: Mapping[str, Optional[int]] = field(default_factory=dict)
+    floats: Mapping[str, Optional[float]] = field(default_factory=dict)
+    dates: Mapping[str, Optional[date]] = field(default_factory=dict)
+    datetimes: Mapping[str, Optional[datetime]] = field(default_factory=dict)
```

Comment:
Having one field per type is verbose but it provides us one (big in my opinion) advantage. The class that uses the result does not need to verify anymore the type of a matches parameter before using it in most cases.
When you get a MatchResult you can just get the specific parameter by name (it must be there if the match was positive) and mypy already knows it has the right type.

The only exception is when we return entire expressions are parameters. In those cases we return the parent class but. we could break this down even more (haivng a columns, function, etc. attributes) if it will be useful.

Reply:
Couldn't the same behavior be implemented by having a single map containing a union of all these types, and providing accessor methods that do the type assertions instead? It seems to be this behavior would make more sense when merging maps with duplicate keys of different types and simplify the merging process.

Reply:
I think it would work. Good idea.